File: trip-kit-booking/src/main/java/com/skedgo/android/tripkit/booking/BookingServiceImpl.java
package com.skedgo.android.tripkit.booking;

import android.support.annotation.VisibleForTesting;

import com.google.gson.Gson;

import java.io.IOException;

import retrofit2.Response;
import rx.Observable;
import rx.Single;
import rx.functions.Func1;

public class BookingServiceImpl implements BookingService {

  private final BookingApi bookingApi;
  private final Gson gson;
  @VisibleForTesting Func1<Response<BookingForm>, Observable<BookingForm>> handleBookingResponse =
      new Func1<Response<BookingForm>, Observable<BookingForm>>() {
        @Override
        public Observable<BookingForm> call(final Response<BookingForm> bookingFormResponse) {

          if (!bookingFormResponse.isSuccessful()) {
            try {
              BookingError e = asBookingError(bookingFormResponse.errorBody().string());
              return Observable.error(e);
            } catch (IOException e) {
              return Observable.error(e);
            }
          }
          return Observable.just(bookingFormResponse.body());
        }
      };

  public BookingServiceImpl(BookingApi bookingApi, Gson gson) {
    this.bookingApi = bookingApi;
    this.gson = gson;
  }

  @Override public Observable<BookingForm> getFormAsync(String url) {
    return bookingApi.getFormAsync(url).flatMap(handleBookingResponse);
  }

  @Override public Observable<BookingForm> postFormAsync(String url, InputForm inputForm) {

    return bookingApi.postFormAsync(url, inputForm)
        .flatMap(handleBookingResponse);
  }

  @VisibleForTesting BookingError asBookingError(String bookingErrorJson) {
    return gson.fromJson(bookingErrorJson, BookingError.class);
  }
}

File: trip-kit-booking/src/main/java/com/skedgo/android/tripkit/booking/BookingServiceImpl.kt
package com.skedgo.android.tripkit.booking

import com.google.gson.Gson
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import retrofit2.Response
import io.reactivex.Observable
import io.reactivex.functions.Function
import org.intellij.lang.annotations.Flow
import java.io.IOException

class BookingServiceImpl(private val bookingApi: BookingApi, private val gson: Gson) : BookingService {
  var handleBookingResponse: Function<Response<BookingForm>, Observable<BookingForm>> =
      Function { bookingFormResponse ->
        if (!bookingFormResponse.isSuccessful) {
          try {
            val e = asBookingError(bookingFormResponse.errorBody()!!.string())
            return@Function Observable.error(e)
          } catch (e: IOException) {
            return@Function Observable.error(e)
          }

        } else {
          if (bookingFormResponse.code() == 204) {
            Observable.just(NullBookingForm as BookingForm)
          } else {
            Observable.just(bookingFormResponse.body()!!)
          }
        }
      }

  override fun getFormAsync(url: String): Flowable<BookingForm> {
    return bookingApi.getFormAsync(url).flatMap(handleBookingResponse).toFlowable(BackpressureStrategy.BUFFER)
  }

  override fun postFormAsync(url: String, inputForm: InputForm): Flowable<BookingForm> =
      bookingApi.postFormAsync(url, inputForm)
          .flatMap(handleBookingResponse).toFlowable(BackpressureStrategy.BUFFER)

  fun asBookingError(bookingErrorJson: String): BookingError {
    return gson.fromJson(bookingErrorJson, BookingError::class.java)
  }
}
